keps/sig-node/753-sidecar-containers/README.md

```diff
@@ -843,24 +843,53 @@ examples:
+the desired behaviors when Pods with sidecars are terminated. While original approach works better
+with truly graceful termination where consistency is more important than time taken, proposed approach
+works for that scenario as well as a more and more popular scenario of limited time to terminate when
+graceful termination is set by external requirement and Pods needs to do best to gracefully terminate
+as much as possible (think of a Spot Instances with 30 seconds notification).
+
+Here is the proposed approach:
+1. Sidecar containers that have a `PreStop` hook will be notified when the Pod has begun terminating
+   by executing the `PreStop` hook. This happens at the same time as regular containers, and begins
+   the Pod's termination grace period countdown.
+2. Sidecar containers enter the `Terminated` state and are no longer restarted if they fail.
+3. Once the last primary container terminates, the last started sidecar container is notified by
+   sending a `SIGTERM` signal, even if `PreStop` hook hasn't completed yet.
+4. The next sidecar (in reverse order) is notified by sending a `SIGTERM` signal after the previous
+   sidecar container terminates.
+5. This continues until all sidecar containers have terminated, or the Pod's termination grace period
+   expires. In the latter case, all containers are terminated with minimum grace period and the Pod
+   will be terminated after that.
+
+If we compare this to the initial proposal, the following behaviors are preserved:
 - Sidecars should not begin termination until all primary containers have
   terminated.
-  - Implicit in this is that sidecars should continue to be restarted until all
-    primary containers have terminated.
 - Sidecars should terminate serially and in reverse order. I.e. the first
   sidecar to initialize should be the last sidecar to terminate.
 
-To address these needs, before promoting this enhancement to Beta, we introduce
-additional termination grace period fields to manage termination duration
-([draft proposal](https://docs.google.com/document/d/1B01EdgWJAfkT3l6CIwNwoskQ71eyuwet3mjiQrMQbU8))
-and leverage these fields to add reverse order termination of sidecar containers
-after the primary containers terminate.
+The additional benefits of this approach comparing to initial proposal:
+- If graceful termination period is short, and mostly taken by the main container, the sidecar containers
+  has more time to gracefully terminate, for example, clear up buffers of logging container.
+- There is absolutely no change in behavior of main containers - they start graceful termination at exact
+  same time as before and can utilize as much of the graceful termination period as they need. The Pod graceful
+  termination period semantic also stay unchanged.
+
+However, the following behaviors is not preserved:
+- Sidecars should continue to be restarted until all primary containers have terminated.
```

Comment:
@howardjohn wdyt from istio perspective? sidecars will NOT be restarted while the main container is in preStop or ignoring the `SIGTERM`

Reply:
Not too worried about this; it is an extreme edge case for the sidecar to crash, and since its during shutdown it doesn't seem so bad

Reply:
I agree this is an edge case.  I didn't follow the motivation for not restarting them here though.  The crashes should be very rare, so wouldn't negatively impact expected shutdown duration. Is there another reason to not restart them?

Reply:
Sidecar containers would have their preStop called upon pod termination beginning and we discussed that it seemed odd to to later retstart a pod after it had already received the preStop.  You couldn't reliably use the preStop as an indicator to shutdown and would be forced to wait for the SIGTERM.

Reply:
@markmandel
I apologize for bringing you into this discussion.

> sidecars will NOT be restarted if crashed while the main container is in preStop or ignoring the SIGTERM

In [Agones](https://agones.dev/site/), we often set the `terminationGracePeriodSeconds` of GameServers to a higher value to prevent Pods from terminating before a game match has finished. Once a game match has finished, the main container (e.g. Unity Server) initiates a graceful termination process and invokes the [Shutdown()](https://agones.dev/site/docs/guides/client-sdks/#shutdown) method via the Agones SDK sidecar container to change the GameServer’s state to `Shutdown`. The duration of these GameServers’s `terminationGracePeriodSeconds` depends on the game’s specifications, but it can sometimes become extended to one hour or more due to multiple game matches being played by the same users. While I understand that this is an edge case, but there is a possibility that the Agones SDK sidecar container may crash. In such scenarios, we expect the sidecar container to be restarted because the main container can no longer change the state of GameServer via Agones SDK sidecar container. Alternatively, the main container cannot invoke the [Health()](https://agones.dev/site/docs/guides/client-sdks/#health) method to notify that the GameServer’s state is heathy leading to an unexpected restart by Agones controller. Could this potentially be an issue for Agones use cases involving sidecar containers? (I’m not sure but It may also present problems for [Dapr](https://docs.dapr.io/) use cases.)

Reply:
Sounds like this is a resolved conversation.

(a) Thanks for pulling me in, I wasn't aware of this KEP (very excited by it! This is going to remove a bunch of hackery we have in Agones) (b) Yes, you raise a really good point -- if the sidecar did crash during the `terminationGracePeriodSeconds` period, we (Agones) would definitely want the sidecar container to restart, so that our connected SDK operations would be able to continue to operate (Health checking being one of them, but also being able to just access all the information that the SDK still provides.

So the line that I read above:

> Sidecars should continue to be restarted until all primary containers have terminated.

Sounds perfect to me.

Reply:
IIUC this is the same argument as https://github.com/kubernetes/enhancements/pull/4183#discussion_r1341703704 ?  @markmandel can you cross-check me, I think we are saying the same thing.

Reply:
Ah! Thanks for the correction. I did not read the top line. So yes - we would definitely prefer it if sidecars would continue to be restarted until the main container have terminated.

And yes, I think we are saying the same thing here. It's exactly the same issue with a logging sidecar.